### PR TITLE
chai, chai-dom, chai-jquery: fix signature of match()/matches() methods

### PR DESCRIPTION
**Pull request by smhxx**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:

**chai:** [API](http://chaijs.com/api/bdd/#method_match) | [Source](https://github.com/chaijs/chai/blob/9a6610eb1e6b6da1645e9aee26ba452c227aa1dc/chai.js#L2422)
**chai-dom:** [API](https://github.com/nathanboktae/chai-dom#matchselector) | [Source](https://github.com/nathanboktae/chai-dom/blob/master/chai-dom.js#L232)
**chai-jquery:** [API](https://github.com/chaijs/chai-jquery#matchselector) | [Source](https://github.com/chaijs/chai-jquery/blob/master/chai-jquery.js#L194)

- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

This PR was previously submitted as #23073 but I rebased after adding new commits (after I closed it,) so GitHub won't allow me to re-open. Anyway, here's the recap, because it took me a while to iron out the kinks in the interactions between the various `@types/chai-xxx` definitions.

The original reason I opened the first PR was that I noticed `match()` and `matches()` (in the BDD-style expect/should interface) are declared as taking either a `RegExp` or a `string` as their parameter; in reality, the parameter *must* be a `RegExp` and a runtime error will be thrown if a string is given. See below:

Error message

That fix is simple enough. There are also two other patches included in this PR which affect the `chai-dom` and `chai-jquery` definitions. Both of these packages provide their own implementations of `match()` which *do* accept strings, but were previously not declared separately because they didn't need to be. Changing the signature of `match()` to *not* accept strings, as is the case with normal, un-extended chai, breaks these two packages' tests, so they have been modified to merge their own overloads of the method into the `Match` interface.

### CODE REVIEW

File: types/chai-dom/chai-dom-tests.ts [hunk 1]
```diff
@@ -8,7 +8,7 @@ function test() {
 
     var testElement = '<div></div>';
     expect(testElement).to.have.attribute('foo', 'bar');
-    expect(testElement).to.have.attr('foo').match(/bar/);
+    expect(testElement).to.have.attr('foo').match('bar');
```
Comment (micksatana) [inline types/chai-dom/chai-dom-tests.ts]: Assertion in chai-dom using `/bar/` in their example. Could you explain why this change to string?

http://chaijs.com/plugins/chai-dom/

File: types/chai-jquery/chai-jquery-tests.ts [hunk 2]
```diff
@@ -24,7 +24,7 @@ function test_css() {
 function test_data() {
     expect($('#foo')).to.have.data('toggle');
     expect($('#foo')).to.have.css('toggle', 'true');
-    expect($('body')).to.have.css('font-family').and.match(/sans-serif/);
+    expect($('body')).to.have.css('font-family').and.match('sans-serif');
```
Comment (micksatana) [inline types/chai-jquery/chai-jquery-tests.ts]: Similarly to above question. Assertion in chai-dom using /sans-serif/ in their example. Could you explain why this change to string?

http://chaijs.com/plugins/chai-jquery/

**Review by micksatana — CHANGES_REQUESTED** [review]
Change in `types/chai/index.d.ts` seem correct according to [chai api doc, match method](http://chaijs.com/api/bdd/#method_match)

However, IMHO, assertions in `types/chai-dom/chai-dom-tests.ts` and `types/chai-dom/chai-jquery-tests.ts` should still able to test both regexp `/bar/` and selector string. So if put back the test cases and passed, then this can be approved.

File: types/chai-dom/chai-dom-tests.ts (hunk 1 above)
Comment (smhxx) [reply types/chai-dom/chai-dom-tests.ts]: That seems to be a mistake, as does the other change you pointed out. I thought I had reverted those changes once I figured out what was going on, but it appears I didn't. I agree that those parameters should still be RegExp since they're testing for a string match (the default chai behavior.) I'll amend my previous commits to reflect those fixes. Thanks for pointing this out.

**Comment by micksatana** [comment]
ps. just my 2-cent, 

in both `chai-dom` and `chai-jquery` you may need to override `interface Match` to accept both `RegExp|string` and use more generic parameter name than `selector` (because it can accept both selector and RegExp).

**Comment by smhxx** [comment]
As I understand it, the `Chai.Match` interfaces declared in `chai-dom` and `chai-jquery` should *merge into* the default `Chai.Match` interface, effectively functioning as a standard method overload (meaning that there's no need to "re-declare" the existing RegExp-based `match()` method.) The final, merged declaration would be:

```typescript
interface Match {
  (regexp: RegExp, message?: string): Assertion; // from @types/chai
  (selector: string): Assertion; // from @types/chai-dom or @types/chai-jquery
}
```

meaning that TypeScript would understand that `match()` can take either of those forms. That seems to be supported by the fact that the (now fixed) tests for `@types/chai-dom` pass despite using both signatures.

**Comment by micksatana** [comment]
Agree. Look great!

**Comment by smhxx** [comment]
(Just redoing my amend because I realized that my editor touched `chai-dom-tests.ts` and added a newline to the end despite there not being any other changes to the file.)

File: types/chai-jquery/index.d.ts [hunk 3]
```diff
@@ -27,6 +27,10 @@ declare namespace Chai {
         disabled(): Assertion;
         (selector: string): Assertion;
     }
+
```
Comment (micksatana) [inline types/chai-jquery/index.d.ts]: According to your suggestion, here should be...
```
interface Match {
    (regexp: RegExp, message?: string): Assertion;
    (selector: string): Assertion;
}
```
ps. I commented here to make sure the bot doesn't think that we don't have conclusion yet (it still have "Awaiting reviewer.." label on the PR).

File: types/chai-dom/index.d.ts [hunk 4]
```diff
@@ -33,6 +45,12 @@ declare namespace Chai {
 
     }
 
+    interface Match {
```
Comment (micksatana) [inline types/chai-dom/index.d.ts]: According to your suggestion, here should be...
```
interface Match {
    (regexp: RegExp, message?: string): Assertion;
    (selector: string): Assertion;
}
```

**Review by micksatana — CHANGES_REQUESTED** [review]

**Comment by smhxx** [comment]
```typescript
namespace Chai {
  interface Match {
    (regexp: RegExp, message?: string): Assertion;
  }
}
```

is already declared in `@types/chai`, so that overload doesn't need to be re-defined. That is, `@types/chai-dom` and `@types/chai-jquery` use [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) to *augment* (rather than override) the existing `Chai` namespace. Likewise, the `Chai.Match` interface merges into the original `Chai.Match`, resulting in the (theoretical) interface I was referring to above.

In other words, `@types/chai` says:

1. A `Match` is something which can be called like a function, taking a RegExp and optional string as parameters, and returning an `Assertion`.
2. The `Assertion` interface has members `match` and `matches`, which conform to the `Match` interface.

Then, `@types/chai-dom` and `@types/jquery` further say:

1. The `Match` interface inside the `Chai` namespace (the same one previously declared) is something which can **also** be called like a function taking *only a string* and returning an `Assertion`.

The end result is that `match()` and `matches()` have two separate signatures, just like normal method overloading except that it occurs across files. If you look at [chai-dom-tests.ts](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/chai-dom/chai-dom-tests.ts#L7), you'll see that it uses *both* signatures of `match()` without error; the normal `@types/chai` form on line 11 to compare a string to a RegExp, and the extended `@types/chai-dom` form on line 22 to compare an `HTMLElement` to a selector string.

**Review by micksatana — APPROVED** [review]
